File: video-processor-service/src/whisperx-alignment.js
/**
 * WhisperX Forced Alignment Wrapper
 * Uses WhisperX API service (cloud) or local Python script for accurate word-level timestamps
 * Takes existing transcription (from ElevenLabs) and audio, returns accurate timestamps
 */

const { exec } = require('child_process');
const { promisify } = require('util');
const { existsSync, createReadStream } = require('fs');
const { join } = require('path');
const FormData = require('form-data');
const axios = require('axios');

const execAsync = promisify(exec);

/**
 * Use WhisperX forced alignment to get accurate timestamps from existing transcription
 * Tries cloud API first, falls back to local Python script
 * 
 * @param {string} audioFile - Path to audio file
 * @param {string} transcriptionText - Existing transcription text (high quality from ElevenLabs)
 * @param {string} language - Language code (default: "ps" for Pashto)
 * @returns {Promise<{text: string, words: Array, segments: Array}>}
 */
async function alignWithWhisperX(audioFile, transcriptionText, language = 'ps') {
  console.log(`🔗 Using WhisperX forced alignment for accurate timestamps...`);
  console.log(`   Audio: ${audioFile}`);
  console.log(`   Text length: ${transcriptionText.length} chars`);
  
  // Check if WhisperX API URL is configured (cloud service)
  const whisperXApiUrl = process.env.WHISPERX_API_URL;
  
  if (whisperXApiUrl) {
    // Use cloud WhisperX API service
    try {
      console.log(`   Using WhisperX API service: ${whisperXApiUrl}`);
      
      const formData = new FormData();
      formData.append('audio', createReadStream(audioFile));
      formData.append('transcription', transcriptionText);
      formData.append('language', language);
      
      const response = await axios.post(`${whisperXApiUrl}/align`, formData, {
        headers: {
          ...formData.getHeaders(),
        },
        maxContentLength: Infinity,
        maxBodyLength: Infinity,
        timeout: 600000, // 10 minutes timeout
      });
      
      const result = response.data;
      
      console.log(`✅ WhisperX API alignment completed:`);
      console.log(`   Words with timestamps: ${result.words?.length || 0}`);
      console.log(`   Segments: ${result.segments?.length || 0}`);
      
      if (result.segments && result.segments.length > 0) {
        console.log(`   First segment: ${result.segments[0].start}s - ${result.segments[0].end}s`);
        console.log(`   Last segment: ${result.segments[result.segments.length - 1].start}s - ${result.segments[result.segments.length - 1].end}s`);
      }
      
      return result;
      
    } catch (error) {
      console.warn(`⚠️ WhisperX API failed: ${error.message}`);
      console.log(`   Falling back to local WhisperX...`);
      // Fall through to local method
    }
  }
  
  // Fallback: Use local Python script
  const scriptPath = join(__dirname, 'whisperx-alignment.py');
  if (!existsSync(scriptPath)) {
    throw new Error(`WhisperX alignment script not found: ${scriptPath}. Set WHISPERX_API_URL for cloud service.`);
  }
  
  // Check if audio file exists
  if (!existsSync(audioFile)) {
    throw new Error(`Audio file not found: ${audioFile}`);
  }
  
  try {
    // Escape transcription text for shell
    const escapedText = transcriptionText.replace(/'/g, "'\\''").replace(/"/g, '\\"');
    
    // Run WhisperX Python script
    const command = `python3 "${scriptPath}" "${audioFile}" "${escapedText}" --language ${language}`;
    console.log(`   Running: python3 whisperx-alignment.py`);
    
    const { stdout, stderr } = await execAsync(command, {
      maxBuffer: 10 * 1024 * 1024, // 10MB buffer
      timeout: 600000, // 10 minutes timeout
    });
    
    if (stderr && !stderr.includes('WARNING')) {
      console.warn(`⚠️ WhisperX stderr: ${stderr}`);
    }
    
    // Parse JSON output
    const result = JSON.parse(stdout.trim());
    
    console.log(`✅ WhisperX alignment completed:`);
    console.log(`   Words with timestamps: ${result.words?.length || 0}`);
    console.log(`   Segments: ${result.segments?.length || 0}`);
    
    if (result.segments && result.segments.length > 0) {
      console.log(`   First segment: ${result.segments[0].start}s - ${result.segments[0].end}s`);
      console.log(`   Last segment: ${result.segments[result.segments.length - 1].start}s - ${result.segments[result.segments.length - 1].end}s`);
    }
    
    return result;
    
  } catch (error) {
    if (error.message.includes('whisperx not installed')) {
      throw new Error('WhisperX not installed. Install with: pip install whisperx OR set WHISPERX_API_URL for cloud service.');
    }
    
    if (error.message.includes('JSON')) {
      console.error(`❌ WhisperX output (first 500 chars):`, stdout?.substring(0, 500));
      throw new Error(`Failed to parse WhisperX output: ${error.message}`);
    }
    
    throw error;
  }
}

module.exports = {
  alignWithWhisperX,
};

File: video-processor-service/src/whisperx-alignment.py
#!/usr/bin/env python3
"""
WhisperX Forced Alignment for accurate word-level timestamps
Takes existing transcription text and audio, returns accurate timestamps
"""

import json
import sys
import argparse
from pathlib import Path

try:
    import whisperx
    import torch
except ImportError:
    print("ERROR: whisperx not installed. Run: pip install whisperx", file=sys.stderr)
    sys.exit(1)


def align_transcription(audio_path: str, transcription_text: str, language: str = "ps", device: str = "cpu"):
    """
    Use WhisperX forced alignment to get accurate word-level timestamps
    
    Args:
        audio_path: Path to audio file
        transcription_text: Existing transcription text (high quality, from ElevenLabs)
        language: Language code (default: "ps" for Pashto)
        device: Device to use ("cpu" or "cuda")
    
    Returns:
        Dictionary with aligned segments and words
    """
    try:
        # Load audio
        audio = whisperx.load_audio(audio_path)
        
        # Load alignment model (wav2vec2 for forced alignment)
        # WhisperX will use a phoneme-based ASR model for alignment
        model_a, metadata = whisperx.load_align_model(language_code=language, device=device)
        
        # Split transcription into sentences
        sentences = [s.strip() for s in transcription_text.split('.') if s.strip()]
        if not sentences:
            sentences = [transcription_text.strip()]
        
        # Align each sentence
        all_words = []
        all_segments = []
        current_time = 0.0
        
        for sentence in sentences:
            if not sentence:
                continue
                
            # Align sentence
            result = whisperx.align([sentence], model_a, metadata, audio, device=device, return_char_alignments=False)
            
            if result and "segments" in result and len(result["segments"]) > 0:
                segment = result["segments"][0]
                words = segment.get("words", [])
                
                if words:
                    # Extract word timestamps
                    segment_words = []
                    for word in words:
                        if "start" in word and "end" in word and "word" in word:
                            segment_words.append({
                                "word": word["word"],
                                "start": word["start"],
                                "end": word["end"]
                            })
                    
                    if segment_words:
                        all_words.extend(segment_words)
                        all_segments.append({
                            "text": sentence,
                            "start": segment_words[0]["start"],
                            "end": segment_words[-1]["end"],
                            "words": segment_words
                        })
        
        return {
            "text": transcription_text,
            "words": all_words,
            "segments": all_segments
        }
        
    except Exception as e:
        print(f"ERROR: WhisperX alignment failed: {str(e)}", file=sys.stderr)
        return None


def main():
    parser = argparse.ArgumentParser(description="WhisperX forced alignment for accurate timestamps")
    parser.add_argument("audio_path", help="Path to audio file")
    parser.add_argument("transcription_text", help="Existing transcription text")
    parser.add_argument("--language", default="ps", help="Language code (default: ps)")
    parser.add_argument("--device", default="cpu", help="Device: cpu or cuda (default: cpu)")
    
    args = parser.parse_args()
    
    # Check if audio file exists
    if not Path(args.audio_path).exists():
        print(f"ERROR: Audio file not found: {args.audio_path}", file=sys.stderr)
        sys.exit(1)
    
    # Perform alignment
    result = align_transcription(args.audio_path, args.transcription_text, args.language, args.device)
    
    if result:
        print(json.dumps(result, ensure_ascii=False, indent=2))
    else:
        print("ERROR: Alignment failed", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

